src/renderer/renderElement.ts

```diff
@@ -1,12 +1,66 @@
-import { ExcalidrawElement } from "../element/types";
+import { ExcalidrawElement, ExcalidrawTextElement } from "../element/types";
 import { isTextElement } from "../element/typeChecks";
-import { getDiamondPoints, getArrowPoints } from "../element/bounds";
+import {
+  getDiamondPoints,
+  getArrowPoints,
+  getElementAbsoluteCoords,
+} from "../element/bounds";
 import { RoughCanvas } from "roughjs/bin/canvas";
 import { Drawable } from "roughjs/bin/core";
 import { Point } from "roughjs/bin/geometry";
 import { RoughSVG } from "roughjs/bin/svg";
 import { RoughGenerator } from "roughjs/bin/generator";
-import { SVG_NS } from "../utils";
+import { SVG_NS, distance } from "../utils";
+import rough from "roughjs/bin/rough";
+
+function generateElementCanvas(element: ExcalidrawElement) {
+  const canvas = document.createElement("canvas");
+  var context = canvas.getContext("2d")!;
+
+  const isLinear = /\b(arrow|line)\b/.test(element.type);
+
+  if (isLinear) {
+    const [x1, y1, x2, y2] = getElementAbsoluteCoords(element);
+    canvas.width = (distance(x1, x2) + 40) * window.devicePixelRatio;
+    canvas.height = (distance(y1, y2) + 40) * window.devicePixelRatio;
+    element.canvasOffsetX =
+      element.x > x1 ? distance(element.x, x1) * window.devicePixelRatio : 0;
+    element.canvasOffsetY =
+      element.y > y1 ? distance(element.y, y1) * window.devicePixelRatio : 0;
+    context.translate(element.canvasOffsetX, element.canvasOffsetY);
+  } else {
+    canvas.width = (element.width + 40) * window.devicePixelRatio;
+    canvas.height = (element.height + 40) * window.devicePixelRatio;
+  }
+
+  context.translate(20, 20);
+  context.scale(window.devicePixelRatio, window.devicePixelRatio);
+
+  var rc2 = rough.canvas(canvas);
+  context.globalAlpha = element.opacity / 100;
+  if (isLinear) {
+    (element.shape as Drawable[]).forEach(shape => rc2.draw(shape));
+  } else if (isTextElement(element)) {
+    const font = context.font;
+    context.font = element.font;
+    const fillStyle = context.fillStyle;
+    context.fillStyle = element.strokeColor;
+    // Canvas does not support multiline text by default
+    const lines = element.text.replace(/\r\n?/g, "\n").split("\n");
+    const lineHeight = element.height / lines.length;
+    const offset = element.height - element.baseline;
+    for (let i = 0; i < lines.length; i++) {
+      context.fillText(lines[i], 0, (i + 1) * lineHeight - offset);
+    }
+    context.fillStyle = fillStyle;
+    context.font = font;
+  } else {
+    rc2.draw(element.shape as Drawable);
```

Comment:
we actually still need shape for several cases: 1) foremost for the rough renderer (although it could be discarded right after paint) 2) we use it for `arrow`/`line` bounds/collisions checks.

Reply:
Okay, let's keep it for now then!